src/core/connection_pool.c

```diff
@@ -0,0 +1,350 @@
+            uint32_t Mask = (RssConfig->RssIndirectionTableLength / sizeof(CXPLAT_PROCESSOR_INFO)) - 1;
+            RssProcIndex = RssProcessorCount;
+            CXPLAT_PROCESSOR_INFO* RssProc = &RssConfig->RssIndirectionTable[RssHash & Mask];
+            for (uint32_t j = 0; j < RssProcessorCount; j++) {
+                #pragma prefast(suppress:6385, "SAL doesn't understand this is safe");
+                if (RssProcessors[j].Group == RssProc->Group && RssProcessors[j].Index == RssProc->Index) {
+                    RssProcIndex = j;
+                    break;
+                }
+            }
+            CXPLAT_DBG_ASSERT(RssProcIndex < RssProcessorCount);
+            if (ConnectionCounts[RssProcIndex] >= ConnectionsPerProc) {
+                //
+                // This processor already has enough connections on it, so try another port number.
+                //
+                continue;
+            }
+
+            CXPLAT_SOCKET* TestSocket = NULL;
+            CXPLAT_UDP_CONFIG TestConfig = {
+                .LocalAddress = &LocalAddress,
+                .RemoteAddress = &ResolvedRemoteAddress
+            };
+            Status = CxPlatSocketCreateUdp(MsQuicLib.Datapath, &TestConfig, &TestSocket);
+            if (QUIC_FAILED(Status)) {
+                //
+                // This port is already in use. Gotta try another port number.
+                //
+                continue;
+            }
+
+            //
+            // Making it this far means we can create the connection!
+            //
+            CxPlatSocketDelete(TestSocket);
+            break;
+
+        } while (TRUE);
+
+        Status = MsQuicConnectionOpen(Registration, Handler, Context, &Connections[i]);
+        if (QUIC_FAILED(Status)) {
+            QuicTraceLogError(
+                ConnPoolOpenConnection,
+                "[conp] Failed to open connection[%u], 0x%x",
+                i,
+                Status);
+            goto Error;
+        }
+        CreatedConnections++;
+
+        Status = MsQuicSetParam(Connections[i], QUIC_PARAM_CONN_REMOTE_ADDRESS, sizeof(ResolvedRemoteAddress), &ResolvedRemoteAddress);
+        if (QUIC_FAILED(Status)) {
+            QuicTraceLogError(
+                ConnPoolSetRemoteAddress,
+                "[conp] Failed to set remote address on connection[%u], 0x%x",
+                i,
+                Status);
+            goto Error;
+        }
+        Status = MsQuicSetParam(Connections[i], QUIC_PARAM_CONN_LOCAL_ADDRESS, sizeof(LocalAddress), &LocalAddress);
+        if (QUIC_FAILED(Status)) {
+            QuicTraceLogError(
+                ConnPoolSetLocalAddress,
+                "[conp] Failed to set local address on connection[%u], 0x%x",
+                i,
+                Status);
+            goto Error;
+        }
+
+        //
+        // The connection was created successfully, add it to the count for this processor.
+        //
+        ConnectionCounts[RssProcIndex]++;
```

Comment:
You're going to have to call `QuicConnStart` otherwise the port you tried to find that was open might get taken. And, really, you're asking for trouble by using a model of 'try to find an open port, then stop using it, and hope you can get it again'. You should find an initial ephemeral port, and then start creating and starting connections on ports until you've got enough successfully started connections.

Reply:
Indeed, there's a TOCTOU issue there, but the requirements from the customer ask that the connection isn't started yet so they can set their own context (which depends on information that isn't known before the call). 

I'd prefer a model that reserves the ports ahead of creating the connections so they can't be taken, but I don't think there's a cross-platform way to do that